charms/metallb-controller/src/charm.py

```diff
@@ -21,93 +21,84 @@
-            name='config-watcher',
-            namespace=self._stored.namespace,
-            labels={'app': 'metallb'},
-            subject_name='metallb-controller'
-        )
+        self.unit.status = MaintenanceStatus("Fetching image information")
+        try:
+            image_info = self.image.fetch()
+        except OCIImageResourceError:
+            logging.exception('An error occured while fetching the image info')
+            self.unit.status = BlockedStatus("Error fetching image information")
+            return
+
+        if not self._stored.k8s_objects_created:
+            self.unit.status = MaintenanceStatus("Creating supplementary "
+                                                 "Kubernetes objects")
+            utils.create_k8s_objects(self._stored.namespace)
+            self._stored.k8s_objects_created = True
+
+        self.unit.status = MaintenanceStatus("Configuring pod")
+        self.set_pod_spec(image_info)
 
-        self.model.unit.status = ActiveStatus("Ready")
+        self.unit.status = ActiveStatus()
         self._stored.started = True
-        self._stored.configured = True
 
-    def on_remove(self, event):
+    def _on_upgrade(self, event):
+        """Occurs when new charm code or image info is available."""
+        self._stored.started = False
+        self._on_start(event)
+
+    def _on_config_changed(self, event):
+        if self.model.config['protocol'] != 'layer2':
+            self.unit.status = BlockedStatus('Invalid protocol; '
+                                             'only "layer2" currently supported')
+            return
+        if self.model.config['iprange'] != self._stored.iprange:
+            self._stored.started = False
+            self._stored.iprange = self.model.config['iprange']
+            self._on_start(event)
+
+    def _on_remove(self, event):
         """Remove of artifacts created by the K8s API."""
-        self.model.unit.status = MaintenanceStatus("Removing pod")
-        logger.info("Removing artifacts that were created with the k8s API")
-        utils.delete_pod_security_policy_with_api(name='controller')
-        utils.delete_namespaced_role_binding_with_api(
-            name='config-watcher',
-            namespace=self._stored.namespace
-        )
-        utils.delete_namespaced_role_with_api(
-            name='config-watcher',
-            namespace=self._stored.namespace
-        )
-        self.model.unit.status = ActiveStatus("Removing extra config done.")
-        self._stored.configured = False
+        self.unit.status = MaintenanceStatus("Removing supplementary "
+                                             "Kubernetes objects")
+        utils.remove_k8s_objects(self._stored.namespace)
+        self.unit.status = MaintenanceStatus("Removing pod")
         self._stored.started = False
 
-    def set_pod_spec(self):
+    def set_pod_spec(self, image_info):
         """Set pod spec."""
         iprange = self.model.config["iprange"].split(",")
```

Comment:
NB: storing the value was just done to avoid unnecessary restarts of the workload.  If the value were larger, or if there were multiple values which might lead to a change, I'd probably just store a hash instead.  I might go ahead and do that anyway, just for future-proofing.